hs_core/hydroshare/utils.py

```diff
@@ -432,17 +596,64 @@ def resource_file_add_process(resource, files, user, extract_metadata=False, **k
     """
-    ret = ResourceFile.objects.create(content_object=resource,
-                                      resource_file=File(f) if not isinstance(f, UploadedFile) else f)
-    # add format metadata element if necessary
-    file_format_type = get_file_mime_type(f.name)
+    if f:
+        if fed_res_file_name_or_path:
+            ret = ResourceFile.objects.create(content_object=resource,
+                                          resource_file=None,
+                                          fed_resource_file=File(f) if not isinstance(f, UploadedFile) else f)
+        else:
+            ret = ResourceFile.objects.create(content_object=resource,
+                                              resource_file=File(f) if not isinstance(f, UploadedFile) else f,
+                                              fed_resource_file=None)
+        # add format metadata element if necessary
+        file_format_type = get_file_mime_type(f.name)
+    elif fed_res_file_name_or_path:
+        size = get_fed_zone_file_size(fed_res_file_name_or_path)
+        ret = ResourceFile.objects.create(content_object=resource, resource_file=None, fed_resource_file=None,
+                                          fed_resource_file_name_or_path=fed_res_file_name_or_path, fed_resource_file_size=size)
+        if fed_copy is not None:
+            try:
+                from_fname = fed_res_file_name_or_path
+                filename = from_fname.rsplit('/')[-1]
+
+                if resource.resource_federation_path:
+                    to_fname = '{base_path}/{res_id}/data/contents/{file_name}'.format(base_path=resource.resource_federation_path,
+                                                                                   res_id=resource.short_id,
+                                                                                   file_name=filename)
+                    istorage = IrodsStorage('federated')
+                else:
+                    to_fname = '{res_id}/data/contents/{file_name}'.format(res_id=resource.short_id, file_name=filename)
+                    istorage = IrodsStorage()
+                if fed_copy:
+                    istorage.copyFiles(from_fname, to_fname)
+                else:
+                    istorage.moveFile(from_fname, to_fname)
+                # update file path now that file has been copied or moved to HydroShare proxy account space
+                ret.fed_resource_file_name_or_path = 'data/contents/{file_name}'.format(file_name=filename)
+                ret.save()
+            except SessionException as ex:
+                # delete the file added if there is any exception
+                ret.delete()
+                # raise the exception for the calling function to inform the error on the page interface
+                raise SessionException(ex.exitcode, ex.stdout, ex.stderr)
+
+        file_format_type = get_file_mime_type(fed_res_file_name_or_path)
+    else:
+        return None
```

Comment:
Better to raise an exception if the inputs to the function are not valid instead of returning ```None```